File: tests/test_step_compare.c
/*
 * Step-by-step token generation with top-5 logit printing.
 * Compares our C engine output with llama-cpp-python reference.
 */

#include "bitnet.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <float.h>

#ifndef BITNET_SOURCE_DIR
#define BITNET_SOURCE_DIR "."
#endif

#define MAX_TOKENS 64
#define NUM_GENERATE 10
#define TOP_K 5
#define VOCAB_SIZE 73448

typedef struct {
    int token_id;
    float logit;
} token_logit_t;

static int cmp_logit_desc(const void *a, const void *b) {
    float la = ((const token_logit_t *)a)->logit;
    float lb = ((const token_logit_t *)b)->logit;
    if (lb > la) return 1;
    if (lb < la) return -1;
    return 0;
}

int main(void) {
    static const char model_path[] = BITNET_SOURCE_DIR "/models/bitcpm4-1b-tq2_0.gguf";
    bitnet_model_t *model = NULL;
    bitnet_context_t *ctx = NULL;
    int tokens[MAX_TOKENS];
    int n_tokens = 0;
    const float *logits = NULL;
    int i = 0;

    model = bitnet_load_model(model_path);
    if (model == NULL) {
        printf("ERROR: Failed to load model\n");
        return 1;
    }

    ctx = bitnet_create_context(model, MAX_TOKENS);
    if (ctx == NULL) {
        printf("ERROR: Failed to create context\n");
        bitnet_free_model(model);
        return 1;
    }

    /* Tokenize the prompt */
    const char *prompt = "The capital of France is";
    n_tokens = bitnet_tokenize(model, prompt, tokens, MAX_TOKENS);
    if (n_tokens <= 0) {
        printf("ERROR: Failed to tokenize\n");
        bitnet_free_context(ctx);
        bitnet_free_model(model);
        return 1;
    }

    printf("Prompt tokens (%d):", n_tokens);
    for (i = 0; i < n_tokens; ++i) {
        printf(" %d", tokens[i]);
    }
    printf("\n");

    /* Print prompt token texts */
    printf("Prompt token texts:\n");
    for (i = 0; i < n_tokens; ++i) {
        char decoded[256];
        bitnet_decode_token(model, tokens[i], decoded, (int)sizeof(decoded));
        printf("  Token %d: %s\n", tokens[i], decoded);
    }

    /* Run eval on prompt tokens */
    if (bitnet_eval(ctx, tokens, n_tokens) != 0) {
        printf("ERROR: Eval failed\n");
        bitnet_free_context(ctx);
        bitnet_free_model(model);
        return 1;
    }

    /* Generate tokens step by step */
    int generated_tokens[NUM_GENERATE];
    printf("\n");

    for (int step = 0; step < NUM_GENERATE; ++step) {
        logits = bitnet_get_logits(ctx);
        if (logits == NULL) {
            printf("ERROR: No logits at step %d\n", step);
            break;
        }

        /* Find top-K tokens */
        token_logit_t *all_logits = (token_logit_t *)malloc(VOCAB_SIZE * sizeof(token_logit_t));
        if (all_logits == NULL) {
            printf("ERROR: malloc failed\n");
            break;
        }
        for (int j = 0; j < VOCAB_SIZE; ++j) {
            all_logits[j].token_id = j;
            all_logits[j].logit = logits[j];
        }
        qsort(all_logits, VOCAB_SIZE, sizeof(token_logit_t), cmp_logit_desc);

        printf("=== Step %d ===\n", step);
        for (int k = 0; k < TOP_K; ++k) {
            char decoded[256];
            bitnet_decode_token(model, all_logits[k].token_id, decoded, (int)sizeof(decoded));
            printf("  Token %d: logit=%.6f text=%s\n",
                   all_logits[k].token_id, all_logits[k].logit, decoded);
        }

        int next_token = all_logits[0].token_id;
        char sel_decoded[256];
        bitnet_decode_token(model, next_token, sel_decoded, (int)sizeof(sel_decoded));
        printf("  >> Selected: Token %d = %s\n", next_token, sel_decoded);

        generated_tokens[step] = next_token;
        free(all_logits);

        /* Feed back the selected token */
        if (bitnet_eval(ctx, &next_token, 1) != 0) {
            printf("ERROR: Eval failed for token %d\n", next_token);
            break;
        }
    }

    /* Summary */
    printf("\n=== Summary ===\n");
    printf("Prompt: \"%s\"\n", prompt);
    printf("Generated tokens:");
    for (i = 0; i < NUM_GENERATE; ++i) {
        printf(" %d", generated_tokens[i]);
    }
    printf("\n");
    printf("Generated text: ");
    for (i = 0; i < NUM_GENERATE; ++i) {
        char decoded[256];
        bitnet_decode_token(model, generated_tokens[i], decoded, (int)sizeof(decoded));
        printf("%s", decoded);
    }
    printf("\n");

    bitnet_free_context(ctx);
    bitnet_free_model(model);
    return 0;
}

File: tests/test_step_compare.py
"""
Step-by-step token generation comparison using llama-cpp-python (reference).
Generates tokens for "The capital of France is" with greedy sampling,
printing top-5 logits at each generation step.

Uses internal _ctx API to get raw logits directly from llama.cpp.
"""

from llama_cpp import Llama
import numpy as np

MODEL_PATH = "/Users/cuick/workdir/AI/158BitNet/models/bitcpm4-1b-tq2_0.gguf"
VOCAB_SIZE = 73448

llm = Llama(model_path=MODEL_PATH, verbose=False, n_ctx=256, logits=True)

prompt = "The capital of France is"
tokens = llm.tokenize(prompt.encode("utf-8"), add_bos=True)
print(f"Prompt tokens ({len(tokens)}): {list(tokens)}")
print(f"Prompt token texts:")
for t in tokens:
    text = llm.detokenize([t]).decode("utf-8", errors="replace")
    print(f"  Token {t}: {repr(text)}")

# Process prompt tokens
llm.eval(tokens)

# Generate 10 tokens step by step
generated_tokens = []
for step in range(10):
    # Use internal API to get raw logits
    logits_ptr = llm._ctx.get_logits()
    logits = np.ctypeslib.as_array(logits_ptr, shape=(VOCAB_SIZE,)).copy()

    # Get top-5 by logit value
    top_indices = np.argsort(logits)[::-1][:5]
    next_token = int(top_indices[0])

    print(f"\n=== Step {step} ===")
    for idx in top_indices:
        text = llm.detokenize([int(idx)]).decode("utf-8", errors="replace")
        print(f"  Token {idx}: logit={logits[idx]:.6f} text={repr(text)}")

    selected_text = llm.detokenize([next_token]).decode("utf-8", errors="replace")
    print(f"  >> Selected: Token {next_token} = {repr(selected_text)}")

    generated_tokens.append(next_token)
    llm.eval([next_token])

print(f"\n\n=== Summary ===")
print(f"Prompt: {repr(prompt)}")
print(f"Generated tokens: {generated_tokens}")
full_text = llm.detokenize(generated_tokens).decode("utf-8", errors="replace")
print(f"Generated text: {repr(full_text)}")
